File: PythonLeetcode/LeetCodeE/385_MiniParser.py



# 385. Mini Parser

# Given a nested list of integers represented as a string, implement a parser to deserialize it.

# Each element is either an integer, or a list -- whose elements may also be integers or other lists.

# Note: You may assume that the string is well-formed:

# String is non-empty.
# String does not contain white spaces.
# String contains only digits 0-9, [, - ,, ].

# Example 1:

# Given s = "324",

# You should return a NestedInteger object which contains a single integer 324.

# Example 2:

# Given s = "[123,[456,[789]]]",

# Return a NestedInteger object containing a nested list with 2 elements:

# 1. An integer containing value 123.
# 2. A nested list containing two elements:
#    i.  An integer containing value 456.
#    ii. A nested list with one element:
#         a. An integer containing value 789.

# """
# This is the interface that allows for creating nested lists.
# You should not implement it, or speculate about its implementation
# """


class NestedInteger(object):
    def __init__(self, value=None):
        """
        If value is not specified, initializes an empty list.
        Otherwise initializes a single integer equal to value.
        """
        if value:
            self.val_ = int(value)
        self.nestedNI_ = []

    def isInteger(self):
        """
        @return True if this NestedInteger holds a single integer, rather than a nested list.
        :rtype bool
        """
        return len(self.nestedNI_)

    def add(self, elem):
        """
        Set this NestedInteger to hold a nested list and adds a nested integer elem to it.
        :rtype void
        """
        self.nestedNI_.append(elem)        

    def setInteger(self, value):
        """
        Set this NestedInteger to hold a single integer equal to value.
        :rtype void
        """
        self.val_ = value

    def getInteger(self):
        """
        @return the single integer that this NestedInteger holds, if it holds a single integer
        Return None if this NestedInteger holds a nested list
        :rtype int
        """
        return self.val_

    def getList(self):
        """
        @return the nested list that this NestedInteger holds, if it holds a nested list
        Return None if this NestedInteger holds a single integer
        :rtype List[NestedInteger]
        """
        return None if not self.nestedNI_ else self.nestedNI_


class MiniParserdeserialize(object):
    def doit(self, s):
        """
        :type s: str
        :rtype: NestedInteger
        """
        stack = []
        num = ''
        current = None
        for c in s:

            if c.isdigit() or c in ('+', '-'):
                num += c

            elif c == '[':
                stack.append(NestedInteger())

            elif c == ',':
                if num :
                    stack[-1].add(NestedInteger(num))
                num = ''

            elif c == ']':
                current = stack.pop()

                if num:
                    current.add(NestedInteger(num))
                    num = ''

                if stack:
                   stack[-1].add(current)

        if num and not current:
            current = NestedInteger(int(num))

        return current


    def doit1(self, s):
        """
        :type s: str
        :rtype: NestedInteger
        """
        return eval(s)


class Solution:
    def deserialize(self, s):
        """
        :type s: str
        :rtype: NestedInteger
        """
        def nestedInteger(x):
            if isinstance(x, int):
                return NestedInteger(x)
            else:
                lst = NestedInteger()
                for y in x:
                    lst.add(nestedInteger(y))
                return lst
        return nestedInteger(eval(s))


class Solution:
    def deserialize(self, s):
        n, i = self.parse(s, 0)
        
        assert i == len(s)
        
        return n
        
    def parse(self, s, i):
        assert i < len(s)
        
        if s[i].isdigit() or s[i] == '-':
            return self.parseInt(s, i)
        else:
            return self.parseList(s, i)
        
    def parseInt(self, s, i):
        assert i < len(s) and (s[i].isdigit() or s[i] == '-')
        
        sign = 1
        
        if s[i] == '-':
            sign = -1
            i += 1
        
        j = i + 1
        while j < len(s) and s[j].isdigit():
            j += 1
        
        return NestedInteger(sign * int(s[i:j])), j
        
    def parseList(self, s, i):
        assert i < len(s) and s[i] == '['
        
        i += 1
        n = NestedInteger()
        
        while s[i] != ']':
            x, i = self.parse(s, i)
            
            assert i < len(s)
            
            n.add(x.getInteger() if x.isInteger() else x.getList())
            
            if s[i] == ',':
                i += 1
                
        return n, i + 1

class Solution:
    def deserialize(self, s):
        """
        :type s: str
        :rtype: NestedInteger
        """
        if s[0] == '[':
            i, ans = self.parse(s, 1)
        else:
            i, ans = self.parse(s, 0)
        return ans
    
    def parse(self, s, i):
        buf = []
        list_ = NestedInteger()
        while i < len(s):
            if s[i] == '[':
                i, retVal = self.parse(s, i + 1)
                list_.add(retVal)
            elif s[i] == ',':
                if len(buf) > 0:
                    list_.add(int(''.join(buf)))
                    buf = []
            elif s[i] == ']':
                if len(buf) > 0:
                    list_.add(int(''.join(buf)))
                return i, list_
            else:
                buf.append(s[i])
            i = i + 1
        
        if len(buf) > 0:
            list_.setInteger(int(''.join(buf)))
        return i, list_


if __name__=="__main__":

    res = MiniParserdeserialize().doit("[]")

    res = MiniParserdeserialize().doit("[-1]")

    res = MiniParserdeserialize().doit("123")

    res = MiniParserdeserialize().doit("[123,[456,[789]]]")

    res = MiniParserdeserialize().doit("[123,456,[788,799,833],[[]],10,[]]")

    pass

File: cppLeetcode/leetcodeM/385_MiniParser.cpp
/*
385. Mini Parser

Given a nested list of integers represented as a string, implement a parser to deserialize it.

Each element is either an integer, or a list -- whose elements may also be integers or other lists.

Note: You may assume that the string is well-formed:

. String is non-empty.
. String does not contain white spaces.
. String contains only digits 0-9, [, - ,, ].
 

Example 1:

Given s = "324",

You should return a NestedInteger object which contains a single integer 324.
 

Example 2:

Given s = "[123,[456,[789]]]",

Return a NestedInteger object containing a nested list with 2 elements:

1. An integer containing value 123.
2. A nested list containing two elements:
    i.  An integer containing value 456.
    ii. A nested list with one element:
         a. An integer containing value 789.
 

*/
#include <stack>
#include <string>
#include <vector>

using std::string;
using std::vector;
using std::stack;

/**
 * // This is the interface that allows for creating nested lists.
 * // You should not implement it, or speculate about its implementation
 * class NestedInteger {
 *   public:
 *     // Constructor initializes an empty nested list.
 *     NestedInteger();
 *
 *     // Constructor initializes a single integer.
 *     NestedInteger(int value);
 *
 *     // Return true if this NestedInteger holds a single integer, rather than a nested list.
 *     bool isInteger() const;
 *
 *     // Return the single integer that this NestedInteger holds, if it holds a single integer
 *     // The result is undefined if this NestedInteger holds a nested list
 *     int getInteger() const;
 *
 *     // Set this NestedInteger to hold a single integer.
 *     void setInteger(int value);
 *
 *     // Set this NestedInteger to hold a nested list and adds a nested integer to it.
 *     void add(const NestedInteger &ni);
 *
 *     // Return the nested list that this NestedInteger holds, if it holds a nested list
 *     // The result is undefined if this NestedInteger holds a single integer
 *     const vector<NestedInteger> &getList() const;
 * };
 */
class Solution {

    // This is the interface that allows for creating nested lists.
    // You should not implement it, or speculate about its implementation
    class NestedInteger {
        public:
        // Constructor initializes an empty nested list.
        NestedInteger();
    
        // Constructor initializes a single integer.
        NestedInteger(int value);
    
        // Return true if this NestedInteger holds a single integer, rather than a nested list.
        bool isInteger() const;
    
        // Return the single integer that this NestedInteger holds, if it holds a single integer
        // The result is undefined if this NestedInteger holds a nested list
        int getInteger() const;
    
        // Set this NestedInteger to hold a single integer.
        void setInteger(int value);
    
        // Set this NestedInteger to hold a nested list and adds a nested integer to it.
        void add(const NestedInteger &ni);
    
        // Return the nested list that this NestedInteger holds, if it holds a nested list
        // The result is undefined if this NestedInteger holds a single integer
        const vector<NestedInteger> &getList() const;
    };
  
    /*
        385.Mini-Parser
        本题涉及了一种不常见的自定义数据结构NestedInteger，需要好好分析。

        本题一个可以取巧的地方在于：任何一个NestedInteger变量都可以完全看做是一个vector，哪怕这个list里的元素只有1.这样就避免了考虑这个NestedInteger变量可能仅有一个int的麻烦。

        另外，注意到NestedInteger类没有给出清零的操作，所以必须尽量不借助多余的变量来实现栈的操作。

        具体的算法是：

        遇到s[i]=='['时，就要新开了一个NestedInteger的变量。这个变量的本质要是一个List，后续的同级元素都是它的子元素。怎么利用构造函数新开一个类的对象呢？
        Stack.push(NestedInteger());
        这样，这个变量不用命名、不用赋初值就可以直接放入栈里。

        遇到s[i]是数字开头时，把完整的数字段抽取出来后得到val。需要这个val转成NestedInteger并放入当前Stack.top()里成为它的一个元素。
        Stack.top().add(Integer(val));
        遇到s[i]==']'时，说明Stack.top()已经把同一层级的所有元素收罗干净了。它本身整个就要投入自己上一级的NestedInteger中，成为它的一个元素。
        NestedInteger temp=Stack.top();
        Stack.pop();
        Stack.top().add(temp);
        特别注意的是，上述语句的前提条件是 Stack.size()>1
    */
    NestedInteger deserialize(string s) 
    {
        if (s[0]!='[') return NestedInteger(stoi(s));
        
        stack<NestedInteger>Stack;
        
        for (int i=0; i<s.size(); i++)
        {
            if (s[i]=='[')
                Stack.push(NestedInteger());
            else 
            {
                int i0=i;
                while (i<s.size() && s[i]!=',' && s[i]!=']') i++;

                if (i>i0)
                {
                    int num = stoi(s.substr(i0,i-i0));
                    Stack.top().add(NestedInteger(num));
                }
                
                if (s[i]==']' && Stack.size()>1)
                {
                    NestedInteger temp=Stack.top();
                    Stack.pop();
                    Stack.top().add(temp);
                }
            }
        }
        
        return Stack.top();
    }

public:
    NestedInteger deserialize_stack(string s) {
        
        if (s[0]!='[') return NestedInteger(stoi(s));
        
        vector<NestedInteger> st;
        long long num = 0;
        bool isnum = false;
        int sig = 1;
        
        for (int i = 0; i < s.length(); i++) {
            
            if (isdigit(s[i])) {
                isnum = true;
                num = num * 10 + s[i] - '0';
            } else if (s[i] == '-') {
                sig = -1;
            }
            else if (s[i] == '[') {
                st.push_back(NestedInteger());
            } else if (s[i] == ',') {
                if (!isnum) continue;
                st.back().add(NestedInteger(sig*num));
                sig = 1;
                num = 0;
                isnum = false;
            } else if (s[i] == ']') {
                if (isnum)
                    st.back().add(NestedInteger(sig * num));
                if (st.size() == 1) continue;
                sig = 1;
                num = 0;
                auto c = st.back();
                st.pop_back();
                st.back().add(c);
                isnum = false;
            }
        }
        
        return st.back();
    }

    char* a;
    void pa(char* x) {
        std::cout << (int)(x - a) << std::endl;
    }
private:
    NestedInteger xdeserialize(char* &x) {
        if (*x == '[') {
            ++x;
            auto y = NestedInteger();
            if (*x == ']') {
                ++x;
                return y;
            }
            for (--x; *x != ']';)
                y.add(xdeserialize(++x));
            ++x;
            pa(x);
            return y;
        }
        else {
            bool s = false;
            if (*x == '-') {
                ++x;
                s = true;
            }
            int z = 0;
            for (; isdigit(*x); ++x) 
                z = z * 10 + *x - '0';
            if (s)
                z *= -1;
            return NestedInteger(z);
        }
    }
public:
    NestedInteger deserialize(string s) {
        char* x = (char*)s.c_str();
        a = x;
        return xdeserialize(x);
    }
};